src/coreclr/jit/morph.cpp

```diff
@@ -10189,9 +10189,14 @@ GenTree* Compiler::fgOptimizeCastOnAssignment(GenTreeOp* asg)
     if (!effectiveOp1->OperIs(GT_IND, GT_LCL_VAR, GT_LCL_FLD))
         return asg;
 
-    if (effectiveOp1->OperIs(GT_LCL_VAR) &&
-        !lvaGetDesc(effectiveOp1->AsLclVarCommon()->GetLclNum())->lvNormalizeOnLoad())
-        return asg;
+    if (effectiveOp1->OperIs(GT_LCL_VAR))
+    {
+        LclVarDsc* varDsc = lvaGetDesc(effectiveOp1->AsLclVarCommon()->GetLclNum());
+
+        // It is not safe to remove the cast for non-NormalizeOnLoad variables or parameters.
```

Comment:
This should mention the "why" - i. e. that this is a quirk/workaround for VN's handling of NOL and that we know this is in fact not a complete fix.

Reply:
I don't that these issues are with VN's handling of NOLs; I could be wrong - it's very hard to determine.

Reply:
If this issue isn't with VN's handling, then what is it with? We should understand the problem before we work around it.

Reply:
Well - we don't know about other bugs related to this, but if there are, surely we would want to find out what they are as well.

FWIW, example from #85382 looks very much like the VN issue.

Reply:
> We should understand the problem before we work around it.

I tried to describe the issue in the description regarding the `mov` elision.

Reply:
> which means sign/zero-extended instructions never get emitted

This is not a problem by itself - the upper bits are presumed to be undefined. In other words, `STORE_LCL_VAR`s do not have an implicit truncation semantic.

Reply:
> STORE_LCL_VARs do not have an implicit truncation semantic.

So that's interesting. We definitely do this today.

Is there a place where NormalizeOnLoad rules are written down? Perhaps I can include these rules as part of this PR.

Reply:
The rules are conceptually simple (though not written down anywhere AFAIK):

1) All small locals are actually `TYP_INT` locals.
2) NOL locals are such that not all definitions can be controlled by the compiler and so the upper bits can be undefined. For parameters this is the case because of ABI. For struct fields - because of padding. For address-exposed locals - because not all stores are direct.
3) Hence, all NOL uses (unless proven otherwise) are assumed in morph to have undefined upper bits and explicit casts have be inserted to "normalize" them back to conform to IL semantics.

This naturally leads to the conclusion that direct stores need not be normalizing - after all, the uses already must assume undefined bits.

Reply:
I think the problem arises when morph concludes that it can forego normalizing in the following code:

https://github.com/dotnet/runtime/blob/f510c5ab2d8a9b94b84f63ff9d2d8318243e390f/src/coreclr/jit/morph.cpp#L4908-L4918

In this case the subrange assertion is coming from the previous
```scala
     ( 21, 24) [000106] -A--G---R--                         ▌  ASG       ushort
     (  4,  3) [000107] D------N---                         ├──▌  LCL_VAR   ushort V00 arg0         
     ( 16, 20) [000108] ----G------                         └──▌  CAST      int <- ushort <- long
     ( 15, 18) [000109] ----G------                            └──▌  RSH       long  
     (  3,  2) [000110] -----------                               ├──▌  LCL_VAR   long   V03 tmp1         
     (  8, 15) [000111] ----G------                               └──▌  AND       int   
     (  6, 13) [000112] n---G------                                  ├──▌  IND       byte  
     (  3, 10) [000113] H----------                                  │  └──▌  CNS_INT(h) long   0x7ff9d2f507ea static Fseq[s_17]
     (  1,  1) [000114] -----------                                  └──▌  CNS_INT   int    63
```
Then LSRA ends up allocating the cast and `V00` to the same register and the store that morph acted on is elided:
```scala
N085 (  9,  6) [000108] ----G------                  t108 = ▌  CAST      int <- long REG rax <l:$2cb, c:$2cc>
                                                            ┌──▌  t108   int    
N087 (  9,  6) [000107] DA--G--N---                         ▌  STORE_LCL_VAR ushort V00 arg0         d:3 rax REG rax $VN.Void
```

That morph code is quite questionable (and essentially puts NOL variables into very awkward territory, since now the frontend/middleend is suddenly reasoning about when NOLs are being allocated into registers). Not sure how bad it would be to remove.

Reply:
It seems likely that a peephole in the emitter would be able to get a large portion of the benefit of that subrange assertion check without having to complicate the NOL semantics with how stores are handled.

Reply:
I did experiment with adding a peephole to look at redundant `movzx` instructions that would help certain cases.

@jakobbotsch so that opt, I disabled it and https://github.com/dotnet/runtime/issues/85382 did pass, but not https://github.com/dotnet/runtime/issues/85081.

Here is what I am using to test:
```
DOTNET_TieredCompilation=0
DOTNET_JitStressRegs=8
```
```csharp
using System.Runtime.CompilerServices;

public class Program
{
    public static long s_15;
    public static sbyte s_17;
    public static ushort s_21 = 36659;
    public static void Main()
    {
        s_15 = ~1;
        var x = (ushort)0;
        bool vr1 = M40(x);
    }

    public static bool M40(ushort arg0)
    {
        for (int var0 = 0; var0 < 2; var0++)
        {
            arg0 = 65535;
            arg0 &= (ushort)(s_15++ >> s_17);
            arg0 %= s_21;
        }

        System.Console.WriteLine(arg0);

        Test.MainTest();

        return false;
    }
}


public class Test
{
    public class Program
    {
        public static short s_2;

        [MethodImpl(MethodImplOptions.NoInlining)]
        public static void Consume(int x) { }

        [MethodImpl(MethodImplOptions.NoInlining)]
        public static int M8(byte arg0)
        {
            s_2 = 1;
            arg0 = (byte)(-s_2);
            var vr1 = arg0 & arg0;
            Consume(vr1);
            return vr1;
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static int MainTest()
    {
        var x = (byte)1;
        var result = Test.Program.M8(x);
        if (result != 255)
        {
            return 0;
        }
        Console.WriteLine("Success");
        return 100;
    }
}
```

A successful run will print:
```
28876
Success
```

Reply:
So I too took a look at the dump for #85382 and the problem (in my interpretation) is that remorphing invalidates previous assumption as made by and acted on in assertion prop (@jakobbotsch - did you mean this too?):

1) We have `NOL = CAST<short>(...)`.
2) Local assertion prop takes a dependency on this, does not add cast for a downstream use.
3) Remorph invalidates the assumption by removing the cast (or, rather making it into a `CAST<int>` one).

Reply:
> (@jakobbotsch - did you mean this too?):

Yes, except I think the optimization made by local assertion prop makes the model of NOL locals much more complex than it should be (i.e. it means your explanation of NOL locals above is really not the fully story).
Local assertion prop here does not only rely on what it proves by seeing the subrange assertion. It also relies on normalization done by `genUnspillRegIfNeeded`, `GT_LCL_VAR` codegen and args homing to make this work. 
In other words: the assertion only meaningfully describes the upper bits if it just so happens that the local ended up being in a 4-byte home for the time between the assignment that created it and the use.

It's very subtle and hard to reason about this in the face of interacting optimizations like this one (another issue I dug up with an interacting optimization: #66624). So if we could get most of the benefits of this optimization through an emitter peephole, and simplify morph to always normalize, then I think that would be much preferable to the current state of things.

Reply:
FWIW, restricting the fix to parameters is not enough, it also needs to be restricted to struct fields. The following example fails in the same way:
```csharp
public static bool M40()
{
    S s = default;
    for (int var0 = 0; var0 < 2; var0++)
    {
        s.U = 65535;
        s.U &= (ushort)(s_15++ >> s_17);
        s.U %= s_21;
    }

    System.Console.WriteLine(s.U);
    return false;
}

public struct S { public ushort U; }
```